### ISSUE TITLE
Online storage deal is stuck in "StorageDealWaitingForData" state

### ISSUE THREAD

**Issue by rjan90** [filecoin-project/lotus#5752]
**Describe the bug**
When you get two online StorageDeals, and the first storage deal is transferring slowly / gets stalled, and the second StorageDeal finishes transferrinng, it is stuck in the `StorageDealWaitingForData` state

The first deal is slowly transferring / stuck:
```
...7awgmtby  0        StorageDealTransferring                  f3vlvigoem5oqaeu6d7g3i5k2nq2vxtjsf2q5pqpbxkzpnhhaucqo7ej5le6hashmimc7krjgxglrmh7bvfcja  32GiB   0.67410816 FIL            1053294
...ivfffaxe  0        StorageDealWaitingForData                f3shjxrnqvzn6x2dxojjpfplq5hhxww4qdtnfm42wosayutnjhtahpeu3qjcidz6y7kmqjrw6uuahhiua4vqyq  32GiB   0 FIL                     1456268
```
While the second StorageDeal has finished transferring, but are stuck in the `StorageDealWaitingForData`-state
```
lotus-miner data-transfers list --completed | grep fffaxe
1071   Completed  ...U4ePmpys     ...ns5wrp72  N           16.77GiB     ...ng5xo35htaejunngdcyhaqhrwzivfffaxe"}}
```  

**To Reproduce**
1. Get two deals
2. The first is slowly transferring / stuck, while the second one finishes
3. The second deal will not be transferred to a sector until the first storage-deal finished transferring (which might not happen)

**Expected behavior**
Expected the second storagedeal to be ready for sealing, instead of waiting for the "first" storage deal to finish transfer.

**Version (run `lotus version`):**
_Daemon:  1.5.0+mainnet+git.55b85f314+api1.1.0
Local: lotus version 1.5.0+mainnet+git.55b85f314_

**Comment by raulk** [comment]
This appears to be some head-of-line blocking in go-data-transfer / go-fil-markets. @dirkmc – when you get a chance, could you analyze further?

**Comment by raulk** [comment]
Seems related to / duplicate of #5584 -- but this issue seems to provide a better lead.

**Comment by rjan90** [comment]
Adding some more information here, as I experienced another round of deals stuck in "StorageDealWaitingForData":

```
...552eeyia  1597656  StorageDealAwaitingPreCommit             f3qpseg3qlpckug75qqas6jsce4up6kg5bo4yaefwnxo4fwx2n5f7jpkkrzy37jxnshkd6mazxso2rxfoaxska  512MiB  0.01 FIL                  1000000
...lvudm6iy  0        StorageDealWaitingForData                f3qpseg3qlpckug75qqas6jsce4up6kg5bo4yaefwnxo4fwx2n5f7jpkkrzy37jxnshkd6mazxso2rxfoaxska  1GiB    0.02 FIL                  1000000
...j3ibwhoq  0        StorageDealWaitingForData                f3qpseg3qlpckug75qqas6jsce4up6kg5bo4yaefwnxo4fwx2n5f7jpkkrzy37jxnshkd6mazxso2rxfoaxska  1GiB    0.02 FIL                  1000000
...74jet4he  0        StorageDealWaitingForData                f3qpseg3qlpckug75qqas6jsce4up6kg5bo4yaefwnxo4fwx2n5f7jpkkrzy37jxnshkd6mazxso2rxfoaxska  2GiB    0.04 FIL                  1000000
...grs4jqna  0        StorageDealWaitingForData                f3qpseg3qlpckug75qqas6jsce4up6kg5bo4yaefwnxo4fwx2n5f7jpkkrzy37jxnshkd6mazxso2rxfoaxska  256MiB  0.005 FIL                 1000000

```
The first storage deal with `Deal-ID: 1597656`, was originally in PendingPublish, while the others was in `StorageDealWaitingForData`, I pushed the message out on-chain to see if the other deals would finish, but they are still stuck in the state, even after the first deal is sealing. Here are some more logs related to the deals:

_Logs of the deal in StorageDealAwaitingPreCommit_
```
tail -30000 ~/lotusminer.log | grep eeyia
2021-03-09T20:15:58.870+0100	INFO	markets	loggers/loggers.go:20	storage provider event	{"name": "ProviderEventOpen", "proposal CID": "bafyreiespuvqjhyynf6ew5ninlxxfuhb32be6whwgewio7ceqw552eeyia", "state": "StorageDealValidating", "message": ""}
2021-03-09T20:15:58.887+0100	INFO	markets	loggers/loggers.go:20	storage provider event	{"name": "ProviderEventDealDeciding", "proposal CID": "bafyreiespuvqjhyynf6ew5ninlxxfuhb32be6whwgewio7ceqw552eeyia", "state": "StorageDealAcceptWait", "message": ""}
2021-03-09T20:15:58.899+0100	INFO	markets	loggers/loggers.go:20	storage provider event	{"name": "ProviderEventDataRequested", "proposal CID": "bafyreiespuvqjhyynf6ew5ninlxxfuhb32be6whwgewio7ceqw552eeyia", "state": "StorageDealWaitingForData", "message": ""}
2021-03-09T20:15:59.283+0100	INFO	markets	loggers/loggers.go:20	storage provider event	{"name": "ProviderEventDataTransferInitiated", "proposal CID": "bafyreiespuvqjhyynf6ew5ninlxxfuhb32be6whwgewio7ceqw552eeyia", "state": "StorageDealTransferring", "message": ""}
2021-03-09T20:23:23.549+0100	INFO	markets	loggers/loggers.go:20	storage provider event	{"name": "ProviderEventDataTransferRestarted", "proposal CID": "bafyreiespuvqjhyynf6ew5ninlxxfuhb32be6whwgewio7ceqw552eeyia", "state": "StorageDealTransferring", "message": ""}
2021-03-09T20:34:58.841+0100	INFO	markets	loggers/loggers.go:20	storage provider event	{"name": "ProviderEventDataTransferRestarted", "proposal CID": "bafyreiespuvqjhyynf6ew5ninlxxfuhb32be6whwgewio7ceqw552eeyia", "state": "StorageDealTransferring", "message": ""}
2021-03-09T20:40:02.682+0100	INFO	markets	loggers/loggers.go:20	storage provider event	{"name": "ProviderEventDataTransferRestarted", "proposal CID": "bafyreiespuvqjhyynf6ew5ninlxxfuhb32be6whwgewio7ceqw552eeyia", "state": "StorageDealTransferring", "message": ""}
2021-03-09T21:32:22.895+0100	INFO	markets	loggers/loggers.go:20	storage provider event	{"name": "ProviderEventDataTransferCompleted", "proposal CID": "bafyreiespuvqjhyynf6ew5ninlxxfuhb32be6whwgewio7ceqw552eeyia", "state": "StorageDealVerifyData", "message": ""}
2021-03-09T21:32:40.416+0100	INFO	markets	loggers/loggers.go:20	storage provider event	{"name": "ProviderEventVerifiedData", "proposal CID": "bafyreiespuvqjhyynf6ew5ninlxxfuhb32be6whwgewio7ceqw552eeyia", "state": "StorageDealReserveProviderFunds", "message": ""}
2021-03-09T21:32:40.435+0100	INFO	markets	loggers/loggers.go:20	storage provider event	{"name": "ProviderEventFundsReserved", "proposal CID": "bafyreiespuvqjhyynf6ew5ninlxxfuhb32be6whwgewio7ceqw552eeyia", "state": "StorageDealReserveProviderFunds", "message": ""}
2021-03-09T21:32:40.439+0100	INFO	markets	loggers/loggers.go:20	storage provider event	{"name": "ProviderEventFunded", "proposal CID": "bafyreiespuvqjhyynf6ew5ninlxxfuhb32be6whwgewio7ceqw552eeyia", "state": "StorageDealPublish", "message": ""}
2021-03-09T21:43:32.035+0100	INFO	markets	loggers/loggers.go:20	storage provider event	{"name": "ProviderEventDealPublishInitiated", "proposal CID": "bafyreiespuvqjhyynf6ew5ninlxxfuhb32be6whwgewio7ceqw552eeyia", "state": "StorageDealPublishing", "message": ""}
2021-03-09T21:50:00.409+0100	INFO	markets	loggers/loggers.go:20	storage provider event	{"name": "ProviderEventFundsReleased", "proposal CID": "bafyreiespuvqjhyynf6ew5ninlxxfuhb32be6whwgewio7ceqw552eeyia", "state": "StorageDealPublishing", "message": ""}
2021-03-09T21:50:00.412+0100	INFO	markets	loggers/loggers.go:20	storage provider event	{"name": "ProviderEventDealPublished", "proposal CID": "bafyreiespuvqjhyynf6ew5ninlxxfuhb32be6whwgewio7ceqw552eeyia", "state": "StorageDealStaged", "message": ""}
2021-03-09T21:50:03.657+0100	INFO	markets	loggers/loggers.go:20	storage provider event	{"name": "ProviderEventDealHandedOff", "proposal CID": "bafyreiespuvqjhyynf6ew5ninlxxfuhb32be6whwgewio7ceqw552eeyia", "state": "StorageDealAwaitingPreCommit", "message": ""}
```

_Logs of the deal(s) in StorageDealWaitingForData_
```
tail -30000 ~/lotusminer.log | grep dm6iy
2021-03-09T20:26:21.576+0100	INFO	markets	loggers/loggers.go:20	storage provider event	{"name": "ProviderEventOpen", "proposal CID": "bafyreidlrhyw3ykqdstunxu6jeuv5hha7rctsodjwtn7js3oyhlvudm6iy", "state": "StorageDealValidating", "message": ""}
2021-03-09T20:26:21.592+0100	INFO	markets	loggers/loggers.go:20	storage provider event	{"name": "ProviderEventDealDeciding", "proposal CID": "bafyreidlrhyw3ykqdstunxu6jeuv5hha7rctsodjwtn7js3oyhlvudm6iy", "state": "StorageDealAcceptWait", "message": ""}
2021-03-09T20:26:21.606+0100	INFO	markets	loggers/loggers.go:20	storage provider event	{"name": "ProviderEventDataRequested", "proposal CID": "bafyreidlrhyw3ykqdstunxu6jeuv5hha7rctsodjwtn7js3oyhlvudm6iy", "state": "StorageDealWaitingForData", "message": ""}
----------------
tail -30000 ~/lotusminer.log | grep whoq
2021-03-09T20:54:59.607+0100	INFO	markets	loggers/loggers.go:20	storage provider event	{"name": "ProviderEventOpen", "proposal CID": "bafyreidiuvoracxcaurd4n7yttzkhpudgqjjqbdfb55lrubdjzj3ibwhoq", "state": "StorageDealValidating", "message": ""}
2021-03-09T20:54:59.619+0100	INFO	markets	loggers/loggers.go:20	storage provider event	{"name": "ProviderEventDealDeciding", "proposal CID": "bafyreidiuvoracxcaurd4n7yttzkhpudgqjjqbdfb55lrubdjzj3ibwhoq", "state": "StorageDealAcceptWait", "message": ""}
2021-03-09T20:54:59.632+0100	INFO	markets	loggers/loggers.go:20	storage provider event	{"name": "ProviderEventDataRequested", "proposal CID": "bafyreidiuvoracxcaurd4n7yttzkhpudgqjjqbdfb55lrubdjzj3ibwhoq", "state": "StorageDealWaitingForData", "message": ""}
```
_What is different in this case is that there are no transfers at all for the deals (completed/failed/ongoing) in StorageDealWaitingForData:_

```
~$ lotus-miner data-transfers list --completed | grep dm6iy
~$
~$ lotus-miner data-transfers list --completed | grep whoq
~$
~$ lotus-miner data-transfers list --show-failed | grep dm6iy
~$
~$ lotus-miner data-transfers list --show-failed | grep whoq
~$
```

**Comment by dirkmc** [comment]
I suspect that the deals are actually stuck on the client side. This could be because of either of the following reasons:
- by default the client only runs a maximum of 6 data transfers concurrently
- there has been a network error and the client has stopped trying to send the deal data

There are a couple of solutions in the works:
- improve performance of the data-transfer module so it can send many deals concurrently
- improve data-transfer module's handling of connection issues

**Comment by cryptowhizzard** [comment]
Hi Dirk, i am the one sending the deal to @rjan90 . Please see the output of my lotus-miner and lotus daemon.

I guess you are right, this problem occurs when there are some transfers stuck.

>>
@lotus-core:~# lotus-miner data-transfers list
Sending Channels

ID   Status           Sending To   Root Cid     Initiated?  Transferred  Voucher                                   
132  Ongoing          ...PX4jGAhC  ...skcayvvc  N           276.1MiB     ...lIncrease":1048576,"UnsealPrice":"0"}  
133  ResponderPaused  ...PX4jGAhC  ...wdrdo246  N           260.1MiB     ...lIncrease":1048576,"UnsealPrice":"0"}  
134  ResponderPaused  ...PX4jGAhC  ...skcayvvc  N           2.505GiB     ...lIncrease":1048576,"UnsealPrice":"0"}  
136  Failing          ...PX4jGAhC  ...wdrdo246  N           300.1MiB     ...lIncrease":1048576,"UnsealPrice":"0"}  
  Message: channel removed due to inactivity
137  ResponderPaused  ...PX4jGAhC  ...skcayvvc  N           309.1MiB     ...lIncrease":1048576,"UnsealPrice":"0"}  
138  Ongoing          ...PX4jGAhC  ...skcayvvc  N           1.08GiB      ...lIncrease":1048576,"UnsealPrice":"0"}  
139  ResponderPaused  ...PX4jGAhC  ...wdrdo246  N           0B           ...lIncrease":1048576,"UnsealPrice":"0"}  

Receiving Channels

ID   Status   Receiving From  Root Cid     Initiated?  Transferred  Voucher                                   
45   Ongoing  ...jRnEvrfc     ...Lvk8mFqd  N           0B           ...dk7t6vpvzk7qea5k2dntbe4nzfamkxqmcm"}}  
52   Ongoing  ...jRnEvrfc     ...z45efNS5  N           0B           ...ztw2nhfwoeiacrfjag4wq3c7lcuqmmkh6m"}}  
61   Ongoing  ...jRnEvrfc     ...Lvk8mFqd  N           0B           ...wacxnbkv5e5drenx2rd4qxexfh6jstojfe"}}  
68   Ongoing  ...jRnEvrfc     ...MXqEEX9Y  N           0B           ...zrvvder44v6ww4adsyew42prdmha5hqrxu"}}  
238  Ongoing  ...jRnEvrfc     ...DF8oTKHJ  N           5.009MiB     ...owa2rhmw7eetscozdlbutusx5b7xp76swu"}}  
278  Ongoing  ...jRnEvrfc     ...1nNKDL9C  N           2.76MiB      ...jysz6qgfi2qlfw2sla7iq5d73pkgwggrt4"}}  
290  Ongoing  ...jRnEvrfc     ...DF8oTKHJ  N           2.258MiB     ...gbsqmilpj2bllpt6sjreknsvufmzymf3ue"}}  
294  Ongoing  ...jRnEvrfc     ...MXqEEX9Y  N           2.509MiB     ...gpw4sa755aotqdyylzvunmdqwr4hvy37ie"}}  
326  Ongoing  ...jRnEvrfc     ...DF8oTKHJ  N           400.6MiB     ...ouhi2wbc62r2kx25oulmbifekgegw64dha"}}  
409  Ongoing  ...jRnEvrfc     ...DF8oTKHJ  N           0B           ...6c7qpgiem5y4eq6tjw4wzgeertfsyk4bme"}}  
435  Ongoing  ...jRnEvrfc     ...ZkCLB8Kq  N           25.02MiB     ...rnyxrwpuc2hlrtq6itf77ix4lljyx3qdji"}}  
446  Ongoing  ...jRnEvrfc     ...ZkCLB8Kq  N           1.404GiB     ...oulcgwe4v2bz46ivemvyntkhwlbjdjrba4"}}  
477  Ongoing  ...jRnEvrfc     ...MXqEEX9Y  N           9.006KiB     ...2i6fqgid5jllpdlwfomyixqxcem4dakpdi"}}  
926  Ongoing  ...U4ePmpys     ...ngxsn6vw  N           0B           ...54peax5lolsblht7gawutrqetxcf6xcmp4"}}  
927  Ongoing  ...U4ePmpys     ...ngxsn6vw  N           0B           ...54peax5lolsblht7gawutrqetxcf6xcmp4"}}  
root@lotus-core:~# lotus client list-transfers
Sending Channels



Receiving Channels

>>>>>

 lotus client list-transfers
Sending Channels



Receiving Channels

**Comment by dirkmc** [comment]
@cryptowhizzard it looks like all the deals you've listed are on the miner side. But you said you're sending deals to @rjan90's miner?

**Comment by cryptowhizzard** [comment]
Correct.

Mar 10 02:17:17  bafyreicbh5hgh2whqvmyjqnjsblbgvr7yh3tpwfwddua3y6u47oc72p4jy  0        f023467   StorageDealCheckForAcceptance  N                 N         baga6ea4seaqlo2jstxop7uatn2sodbvmbj7bivrab5c4rmx563aprch5x7xrcpq  31.75 GiB  0 FIL           1457142   12D3KooWQrktQtwVbMBoeoN6te6i26s1gQPVfTYVgFs1U4ePmpys-12D3KooWMpuRzgVuRS6m285cucGgnXWwTnsNtGzjsMJeEqS7UtJw-1081  Completed       true      Provider state: StorageDealWaitingForData
@lotus-core:~#

**Comment by cryptowhizzard** [comment]
The issue occurs after a few days.... then i have to clear those broken transfers above it seems to get everything going again.

**Comment by cryptowhizzard** [comment]
@dirkmc correct. Seems to be related however. The only way to come out of this "state" is to do a full restart of lotus and the miner , clear / cancel those transfers and empty the datatransfer folder.

**Comment by benjaminh83** [comment]
I'm seeing the same issue: 
```
...krp4l5su  0        StorageDealTransferring                  f3rab4mhsr7f2gdg6ozgj44mtny6bhcy3rvlftc5xyki6t4hvhdkaji3ahfaid6fh2o3wff4yfrzag4lge2r2q  32GiB     0 FIL                     519532
...iqtjdef4  0        StorageDealWaitingForData                f3shjxrnqvzn6x2dxojjpfplq5hhxww4qdtnfm42wosayutnjhtahpeu3qjcidz6y7kmqjrw6uuahhiua4vqyq  16GiB     0 FIL                     1458066
```

```
$ lotus-miner data-transfers list --completed -v | grep bafyreigbotxapo6byz4u6hajdbdx6uiht7mwjiydq7amgzwe73iqtjdef4
1089   Completed   12D3KooWQrktQtwVbMBoeoN6te6i26s1gQPVfTYVgFs1U4ePmpys  bafykbzaced2stiuun52mflabrembmplvzz3r5df4g6jvpizkbxjobw5ap3xkg                                                                                                                                               N           14.92GiB     {"Proposal":{"/":"bafyreigbotxapo6byz4u6hajdbdx6uiht7mwjiydq7amgzwe73iqtjdef4"}} 
```

```
$ lotus-miner data-transfers list --completed -v | grep bafyreihbxpvlo4sh6d6r4vxve3jd3ndnptsuwxwxykxgpy57fhkrp4l5su
1842   Completed   12D3KooWHAN67HgLoqrygQnouFF5uVhwM4kP7zurpR65Powtfgsa  QmNcvtqTqvuskYUYvpjuGsyvibWjjCwaqQTyGxfmtKtg7B                                                                                                                                                               N           28.01GiB     {"Proposal":{"/":"bafyreihbxpvlo4sh6d6r4vxve3jd3ndnptsuwxwxykxgpy57fhkrp4l5
```

**Comment by Reiers** [comment]
I'm seeing the same thing - the transfers is completed, but are stuck in **StorageDealTransferring** state.

```
1889   Completed  ...Powtfgsa     ...nBRS8Av5  N           28.01GiB     ...ht5ewmxiwpu764h3o5hup4emhwygq7ady4"}}  
1885   Completed  ...Powtfgsa     ...2QsNATKC  N           28.01GiB     ...rr6ektomypitjmrx4froje2lnhwr6mshoe"}}  
```

```
...wr6mshoe  0        StorageDealTransferring                  f3rab4mhsr7f2gdg6ozgj44mtny6bhcy3rvlftc5xyki6t4hvhdkaji3ahfaid6fh2o3wff4yfrzag4lge2r2q  32GiB   0 FIL                           523015
...ygq7ady4  0        StorageDealTransferring                  f3rab4mhsr7f2gdg6ozgj44mtny6bhcy3rvlftc5xyki6t4hvhdkaji3ahfaid6fh2o3wff4yfrzag4lge2r2q  32GiB   0 FIL                           522824
```

**Lotus Version:**
```
Daemon:  1.5.2+mainnet+git.3be4984d9+api1.1.0
Local: lotus version 1.5.2+mainnet+git.3be4984d9
```

**Comment by kenshyx** [comment]
I encountered the same problem, on the storage-deals list I have
```
Mar 17 20:56:42  true   bafyreihw2aibm7hfds2do5jnnmhjzf63qpkd5qysnhk2hguojuvcp23x7q  0        StorageDealProviderTransferAwaitRestart  f3rab4mhsr7f2gdg6ozgj44mtny6bhcy3rvlftc5xyki6t4hvhdkaji3ahfaid6fh2o3wff4yfrzag4lge2r2q  32GiB   0 FIL
```
And on the data-transfer I have
`
1851   Completed  12D3KooWHAN67HgLoqrygQnouFF5uVhwM4kP7zurpR65Powtfgsa QmeSqshr6xZGy3CasEWucKQTRh3h3X6wb1eHpBA4temKZy                                                                                                                                                      N      28.01GiB     {"Proposal":{"/":"bafyreihw2aibm7hfds2do5jnnmhjzf63qpkd5qysnhk2hguojuvcp23x7q"}}  
`

**Comment by neondragon** [comment]
+1, and the Ongoing data-transfer says it has transferred much more than expected (73.53GiB)

```
# lotus-miner data-transfers list --completed | grep Powtfgsa
1856   Completed  ...Powtfgsa     ...9aEJE4UQ  N           28.01GiB     ...7bbyf4qg7wc2vipkag6hgs2kpv7hpzprri"}}
1883   Completed  ...Powtfgsa     ...2QsNATKC  N           28.01GiB     ...vzkmq6rmmgqgzmmpihob4jemc6baqpns3e"}}
1901   Ongoing    ...Powtfgsa     ...iGNUZo1C  N           73.53GiB     ...p4si646yjlgtc2wj6ja2rkx2uha4gttw7a"}}
```

```
# lotus-miner storage-deals list | grep 2r2q
...7hpzprri  0        StorageDealTransferring                  f3rab4mhsr7f2gdg6ozgj44mtny6bhcy3rvlftc5xyki6t4hvhdkaji3ahfaid6fh2o3wff4yfrzag4lge2r2q  32GiB   0 FIL                     518641
...baqpns3e  0        StorageDealWaitingForData                f3rab4mhsr7f2gdg6ozgj44mtny6bhcy3rvlftc5xyki6t4hvhdkaji3ahfaid6fh2o3wff4yfrzag4lge2r2q  32GiB   0 FIL                     520729
...a4gttw7a  0        StorageDealWaitingForData                f3rab4mhsr7f2gdg6ozgj44mtny6bhcy3rvlftc5xyki6t4hvhdkaji3ahfaid6fh2o3wff4yfrzag4lge2r2q  32GiB   0 FIL                     520322
```

`lotus-miner version 1.5.2+mainnet+git.cd293aa72.dirty`

**Comment by cryptowhizzard** [comment]
After looking at this issue for weeks , i think i have narrowed it down for you guys.

From a client perspective ( So when i am the one sending out deals ) it seems the deals get stuck in StorageDealWaitingForData when i send out another deal , before the first deal makes it to state publishing. 

I updated scripts tonight , normally speedium sends out deals in N+1 ... so when i tried to move to N+2 this issue came visible when sending out in a faster way.

@dirkmc i hope this helps you.

**Comment by NSC-FIL** [comment]
I'm experiencing the same issues with deals stuck in "StorageDealWaitingForData" state as well, but have another situation with deals in "StorageDealStaged" state that may be related.  Transfer of deals that successfully transferred data should automatically spawn WaitDeals sector(s) to be sealed, however, several of my deals have been stuck for over two days in the following state:
```

...ppjy4oja  1612361  StorageDealStaged                        f3rab4mhsr7f2gdg6ozgj44mtny6bhcy3rvlftc5xyki6t4hvhdkaji3ahfaid6fh2o3wff4yfrzag4lge2r2q  32GiB   0 FIL                     521438
...cfjegt4q  1612451  StorageDealStaged                        f3rab4mhsr7f2gdg6ozgj44mtny6bhcy3rvlftc5xyki6t4hvhdkaji3ahfaid6fh2o3wff4yfrzag4lge2r2q  32GiB   0 FIL                     521045
...ypaetbfq  1612273  StorageDealStaged                        f3sml2wt7uzarci2pjvzra64mapi2hgslhtuwyvxgkxi2dbbazxux5eunb5imdo2v2hyylfr3cpe5qjxvrsj3q  128MiB  0.000650735 FIL           520588
```

```

**Comment by dirkmc** [comment]
My theory is that the storage deals get stuck in `StorageDealWaitingForData` on the provider side because the transfer is blocked on the client side.

When a client makes a storage deal with a miner, the data transfer sequence is:
1. Client sends "open data transfer channel" message: "I want to send file with root CID xyz"
2. Provider makes graphsync request to client: "give me the blocks for CID xyz"
3. Client sends blocks for CID xyz

Currently in graphsync we limit the number of concurrent data transfers to 6. This is to prevent graphsync being overwhelmed by many simultaneous requests.

So I believe that when transfers on the provider are stuck in `StorageDealWaitingForData`, in reality the root cause is the graphsync concurrent transfer limit on the client side - the provider is waiting for the client to respond to the graphsync request, so the provider remains "waiting for data".

The graphsync request from provider to client is authenticated, and with the recent improvements to graphsync memory usage it should be safe to raise the concurrent transfer limit.

I've opened an issue against graphsync to track work on making the concurrent transfer limit configurable: https://github.com/ipfs/go-graphsync/issues/162

Once the limit is configurable, clients can experiment with raising the limit and see if it solves this issue.

**Comment by aarshkshah1992** [comment]
@dirkmc 

- This is already configurable in `Graphsync` and Lotus sets a default limit of 20. Code is [here](https://github.com/filecoin-project/lotus/blob/5d44ed0b237f21e6e8bc7aef0dd90c7590ecde41/node/config/def.go#L209).
- If the users want to increase this limit, they can do so by changing the value of `SimultaneousTransfers ` in the `~/.lotus/config file`.

```
[Client]
#  UseIpfs = false
#  IpfsOnlineMode = false
#  IpfsMAddr = ""
#  IpfsUseForRetrieval = false
#  **SimultaneousTransfers = 20**
````

**Comment by dirkmc** [comment]
@cryptowhizzard does this theory line up with what you're seeing?

Specifically the theory is that the limit is actually on the client side - I would expect that the number of concurrent transfers you can send is equal to the value of `SimultaneousTransfers` in your config file.

Now that we've made improvements to graphsync's memory usage, I would expect you can increase the number of simultaneous transfers. Would you like to try experimenting by increasing `SimultaneousTransfers` and see if you can get higher concurrency?

**Comment by cryptowhizzard** [comment]
This issue seems resolved now since 1.6.0

**Comment by dirkmc** [comment]
@rjan90 are you still seeing this issue?

**Comment by rjan90** [comment]
That’s great! Just updated to v1.6.0, so I have not experienced that many deals yet! Closing the issue for now, will re-open if I experience it again!
